### ISSUE TITLE
Input and store past keepers

### ISSUE THREAD

**Issue by seanconnollydev** [seanconnollydev/yahoo-fantasy-football-tools#44]
This is a multipart story.

An authorized user must enter their league's keeper settings (since these are not available from Yahoo's API).
- Can/cannot keep players kept in a prior season.
- This must happen before any user can view keeper eligibility for the league.
- These settings should be reflected when analyzing keeper eligibility.

An authorized user can enter one or more players kept by each team from a prior season.
- These results will be available for any user authorized to view the league's data.
- Consider building audit trails for these entries.

An authorized user can modify the player-keeper associations from a prior season.
- Any player in the league should be able to modify this data.
- Consider building audit trails for these modifications.

**Comment by seanconnollydev** [comment]
It looks like NHibernate requires a lot of configuration.  I should evaluate Fluent NHibernate:
https://github.com/jagregory/fluent-nhibernate/wiki/Getting-started

**Comment by seanconnollydev** [comment]
Fluent NHibernate looks reasonable enough.  I also need to find a DB migration solution for deployments to prod.  In dev, I think it is reasonable to build up and tear down the full DB locally.

It looks like NHibernate has a limited schema update feature.  A few options are also listed here: http://stackoverflow.com/questions/933327/nhibernate-schemaupdate.

AppHarbor has a post on using NHibernate for automatic migrations: http://blog.appharbor.com/2012/06/11/automatic-migrations-with-nhibernate (Note they migrate on application start).

AppHarbor also has a post on using EF for automatic migrations: http://blog.appharbor.com/2012/04/24/automatic-migrations-with-entity-framework-4-3

Also consider Fluent Migrator: https://github.com/schambers/fluentmigrator

**Comment by seanconnollydev** [comment]
Go here to setup AppHarbor for SQL Server usage: http://support.appharbor.com/kb/add-ons/using-sequelizer

**Comment by seanconnollydev** [comment]
I wasn't able to test this since SQLite was not playing nicely with FluentNHibernate.  I should try this against a local version of SQLExpress, which will more closely resemble AppHarbor's environment.

**Comment by seanconnollydev** [comment]
A few things remaining here:
- Wire up session factory creation in the NInjectControllerFactory
- Create the Edit Keeper Settings page
- Put the analyzers behind an interface so I can write unit tests against them (this may require a separate issue, it requires a fair amount of refactoring).

**Comment by seanconnollydev** [comment]
Next step is to continue work on the Edit Keeper Settings page, using Who Can Help Me as an example for the model binding pattern that I want to try: https://github.com/jongeorge1/Who-Can-Help-Me/blob/master/Solutions/WhoCanHelpMe.Web.Controllers/Profile/ProfileController.cs

After that, I still want to put the analyzers behind an interface.

**Comment by seanconnollydev** [comment]
Another ViewModel -> Entity pattern is to use a builder interface: http://stackoverflow.com/questions/2206005/how-to-map-view-model-back-to-domain-model-in-a-post-action

**Comment by seanconnollydev** [comment]
I still need to wire up editing.  There are some strange interactions between models and domain objects when viewing and editing settings, which is why it isn't working.

I also should look into the ignore list, because I think it ignores my packages folders, which I probably need to include.  I think this is why my references to FluentNHibernate and MvcContrib broke.

**Comment by seanconnollydev** [comment]
Later, consider separating local unit testing database from web application.

**Comment by seanconnollydev** [comment]
Remaining:
- Take the setting into account when checking keeper eligibility
- Unit tests for new action methods
- Add a link to edit keeper settings from the keeper eligibility page
- Put analyzers behind an interface
- Look into ignore list
- Add breadcrumbs to Keeper Settings page
- Look into why leagues list does not populate on first load

**Comment by seanconnollydev** [comment]
Remaining:

Unit tests for new action methods
Put analyzers behind an interface
Look into ignore list
Fix download eligible keeper issue (see TODO)

**Comment by seanconnollydev** [comment]
Remaining:
- At least one more UpdateKeeperSettings unit test required (the positive scenario)
- Add user friendly messages to explain what is happening at different stages (e.g. when forced to enter settings and when settings are saved).
- Put analyzers behind an interface

**Comment by seanconnollydev** [comment]
I think this is ready to merge, so remaining items are:
- Review all changes associated with this feature
- Merge to master (remember to squash)
- Create a new issue for mapping kept players and that supported workflow
- Create a new issue for putting analyzers behind an interface (so they can be unit tested)

**Comment by seanconnollydev** [comment]
Reviewed, tested, and deployed.
